src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DefaultClientNotifierService.cs

```diff
@@ -61,4 +70,12 @@ public override Task OnInitializedAsync(VSInternalClientCapabilities clientCapab
         _initializedCompletionSource.TrySetResult(true);
         return Task.CompletedTask;
     }
+
+    private static ImmutableDictionary<string, object?> GetValues(string method)
+    {
+        return new Dictionary<string, object?>()
+        {
+            {"eventscope.method", method}
+        }.ToImmutableDictionary();
```

Comment:
There's no fast path between a `Dictionary<,>` and an `ImmutableDictionary<,>`, so this is just extra allocations. I'd recommend avoiding the `Dictionary<,>` allocation altogether:

```suggestion
        return ImmutableDictionary<string, object?>.Empty.Add("eventscope.method", method);
```

Reply:
❓does ImmutableDictionary.Add not cause allocations as well?

Reply:
It effectively results in a single allocation.

Reply:
`ToImmutableDictionary()` also results in enumerator allocation too, since it's an extension method for `IEnumerable<KeyValuePair<TKey, TValue>>`. It eventually ends up [here](https://source.dot.net/#System.Collections.Immutable/System/Collections/Immutable/ImmutableDictionary_2.cs,7e63a0690622444f,references).

Ultimately, `new Dictionary<,>.ToImmutableDictionary()` has a lot of extra hidden allocations. :-(

Reply:
🖤 thank you for the detailed explanation